replication-api-impl/src/main/java/org/codice/ditto/replication/api/impl/data/AbstractPersistable.java

```diff
@@ -0,0 +1,81 @@
+/**
+ * Copyright (c) Connexta
+ *
+ * <p>This is free software: you can redistribute it and/or modify it under the terms of the GNU
+ * Lesser General Public License as published by the Free Software Foundation, either version 3 of
+ * the License, or any later version.
+ *
+ * <p>This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+ * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details. A copy of the GNU Lesser General Public
+ * License is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ */
+package org.codice.ditto.replication.api.impl.data;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.UUID;
+import org.codice.ditto.replication.api.data.Persistable;
+
+/**
+ * This class defines some data that any object to be saved in persistence should have and methods
+ * to manipulate that data.
+ */
+public abstract class AbstractPersistable implements Persistable {
+
+  static final String ID_KEY = "id";
+
+  private String id;
+
+  private int version;
+
+  @Override
+  public String getId() {
+    return id;
+  }
+
+  @Override
+  public void setId(String id) {
+    this.id = id;
+  }
+
+  @Override
+  public int getVersion() {
+    return version;
+  }
+
+  public void setVersion(int version) {
+    this.version = version;
+  }
+
+  @Override
+  public void generateId() {
+    this.id = UUID.randomUUID().toString();
+  }
+
+  public abstract String getType();
+
+  /**
+   * Writes the variables of the persistable to a map. Any implementation of this method in a
+   * subclass should first make a call to the super version before performing its own functionality.
+   *
+   * @return a map containing the variable of the persistable
+   */
+  public Map<String, Object> writeToMap() {

```

Comment:
Should writeToMap and readFromMap be methods defined in Persistable? Maybe also think about renaming the methods to something like `toMap` and `fromMap`. The read and write part make it not as clear in my mind.

Reply:
I don't think they should be. We do not want to bleed our implementation to the public API.

Reply:
I can see the arguments for either way so I think I'm going to stick with the easy route of not changing the persistable interface. I will change the names to the ones you suggested however.